lib/clipping.h

```diff
@@ -0,0 +1,229 @@
+            x = p0.x + (p1.x - p0.x) * t;
+            y = p0.y + (p1.y - p0.y) * t;
+            z = box.min.z;
+        }
+
+        if (outcode_out == outcode0) {
+            p0.x = x;
+            p0.y = y;
+            p0.z = z;
+            outcode0 = compute_outcode(p0, box);
+        } else {
+            p1.x = x;
+            p1.y = y;
+            p1.z = z;
+            outcode1 = compute_outcode(p1, box);
+        }
+    }
+
+    return true;
+}
+
+
+enum class PointPlanePos {
+    ON_PLANE, BEHIND_PLANE, IN_FRONT_OF_PLANE
+};
+
+//
+// Classify point to the plane given by equation `n * x = d`.
+//
+inline PointPlanePos classify_point_to_plane(
+    const Vec& pt, const Vec& n, float d)
+{
+    float dist = n * pt - d;
+    if (dist > EPS) {
+        return PointPlanePos::IN_FRONT_OF_PLANE;
+    } else if (dist < -EPS) {
+        return PointPlanePos::BEHIND_PLANE;
+    }
+    return PointPlanePos::ON_PLANE;
+}
+
+//
+// Sutherland-Hodgman polygon clipping at a (thick) plane.
+//
+inline std::vector<Vec>
+clip_polygon_at_plane(const std::vector<Vec> poly, const Vec& n, float d) {
+    assert(poly.size() > 1);
+
+    std::vector<Vec> points;
+
+    Vec a = poly.back();
+    auto a_side = classify_point_to_plane(a, n, d);
+
+    for (auto b : poly) {
+
+        auto b_side = classify_point_to_plane(b, n, d);
+
+        if (b_side == PointPlanePos::IN_FRONT_OF_PLANE) {
+            if (a_side == PointPlanePos::BEHIND_PLANE) {
+                // intersect (a, b) at plane
+                float t;
+                assert(intersect_segment_plane(a, b, n, d, t));
+                Vec pt(a + t * (b - a));
+                assert(
+                    classify_point_to_plane(pt, n, d)
+                    == PointPlanePos::ON_PLANE);
+                points.emplace_back(pt);
+            }
+            points.emplace_back(b);
+        } else if (b_side == PointPlanePos::BEHIND_PLANE) {
+            if (a_side == PointPlanePos::IN_FRONT_OF_PLANE) {
+                // intesect (a, b) at plane
+                float t;
+                assert(intersect_segment_plane(a, b, n, d, t));
+                Vec pt(a + t * (b - a));
+                assert(
+                    classify_point_to_plane(pt, n, d)
+                    == PointPlanePos::ON_PLANE);
+                points.emplace_back(pt);
+            }
+            // a is behind plane or on plane
+        } else {
+            // b is on plane
+            points.emplace_back(b);
+        }
+
+        a = b;
+        a_side = b_side;
+    }
+
+    return points;
+}
+
+
+//
+// Clip triangle `tri` at AABB `box`.
+//
+// Requirement: triangle and box intersect.
+//
+// Return:
+//   the bounding box of the clipped polygon.
+//
+inline Box clip_triangle_at_aabb(const Triangle& tri, const Box& box) {
+    std::vector<Vec> points(tri.vertices.begin(), tri.vertices.end());
+
+    // clip at 6 planes defined by box
+    Axis ax = Axis::X;
+    for (int i = 0; i < 3; ++i, ++ax) {
```

Comment:
Mir gefällt die 3 irgendwie nicht. Hm. kann man nicht irgendwie for(a : Axes) {} machen?